src/stapi_fastapi/routers/product_router.py

```diff
@@ -0,0 +1,124 @@
+
+from stapi_fastapi.constants import TYPE_GEOJSON, TYPE_JSON
+from stapi_fastapi.exceptions import ConstraintsException
+from stapi_fastapi.models.opportunity import (
+    OpportunityCollection,
+    OpportunityRequest,
+)
+from stapi_fastapi.models.order import Order
+from stapi_fastapi.models.product import Product
+from stapi_fastapi.models.shared import Link
+from stapi_fastapi.responses import GeoJSONResponse
+from stapi_fastapi.types.json_schema_model import JsonSchemaModel
+
+if TYPE_CHECKING:
+    from stapi_fastapi.routers import RootRouter
+
+
+class ProductRouter(APIRouter):
+    def __init__(
+        self,
+        product: Product,
+        root_router: RootRouter,
+        *args,
+        **kwargs,
+    ) -> None:
+        super().__init__(*args, **kwargs)
+        self.product = product
+        self.root_router = root_router
+
+        self.add_api_route(
+            path="",
+            endpoint=self.get_product,
+            name=f"{self.root_router.name}:{self.product.id}:get-product",
+            methods=["GET"],
+            summary="Retrieve this product",
+        )
+
+        self.add_api_route(
+            path="/opportunities",
+            endpoint=self.search_opportunities,
+            name=f"{self.root_router.name}:{self.product.id}:search-opportunities",
+            methods=["POST"],
+            response_class=GeoJSONResponse,
+            summary="Search Opportunities for the product",
+        )
+
+        self.add_api_route(
+            path="/constraints",
+            endpoint=self.get_product_constraints,
+            name=f"{self.root_router.name}:{self.product.id}:get-constraints",
+            methods=["GET"],
+            summary="Get constraints for the product",
+        )
+
+        self.add_api_route(
+            path="/order",
+            endpoint=self.create_order,
+            name=f"{self.root_router.name}:{self.product.id}:create-order",
+            methods=["POST"],
+            response_class=GeoJSONResponse,
+            status_code=status.HTTP_201_CREATED,
+            summary="Create an order for the product",
+        )
+
+    def get_product(self, request: Request) -> Product:
+        return self.product.with_links(
+            links=[
+                Link(
+                    href=str(
+                        request.url_for(
+                            f"{self.root_router.name}:{self.product.id}:get-product",
+                        ),
+                    ),
+                    rel="self",
+                    type=TYPE_JSON,
+                ),
+            ],
+        )
+
+    async def search_opportunities(
+        self, search: OpportunityRequest, request: Request
+    ) -> OpportunityCollection:
+        """
+        Explore the opportunities available for a particular set of constraints
+        """
+        try:
+            opportunities = await self.product.backend.search_opportunities(
```

Comment:
We should pass `self` instead of `self.product` here and in the `create_order` method. We'll need to update the `ProductBackend` protocol correspondingly and the test implementation.